File: jscomp/others/bs_HashMultiMap.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
(**  Adapted by Authors of BuckleScript 2017                           *)

module N = Bs_internalBuckets 
module C = Bs_internalBucketsType
module A = Bs_Array

type ('a, 'id) eq = ('a, 'id) Bs_Hash.eq
type ('a, 'id) hash = ('a, 'id) Bs_Hash.hash
type ('a, 'id) dict = ('a, 'id) Bs_Hash.t                        

type ('a, 'b,'id) t = ( ('a, 'id) hash, ('a, 'id) eq ,  'a,'b) N.t


let rec insert_bucket ~hash ~h_buckets ~ndata_tail h old_bucket = 
  match C.toOpt old_bucket with 
  | None -> ()
  | Some cell ->
    let nidx = (Bs_Hash.getHashInternal hash) (N.key cell) [@bs] land (Array.length h_buckets - 1) in 
    let v = C.return cell in 
    begin match C.toOpt (A.getUnsafe ndata_tail nidx) with
      | None -> 
        A.setUnsafe h_buckets nidx  v
      | Some tail ->
        N.nextSet tail v  (* cell put at the end *)            
    end;          
    A.setUnsafe ndata_tail nidx  v;
    insert_bucket ~hash ~h_buckets ~ndata_tail h (N.next cell)


let resize ~hash h =
  let odata = C.buckets h in
  let osize = Array.length odata in
  let nsize = osize * 2 in
  if nsize >= osize then begin (* no overflow *)
    let h_buckets = A.makeUninitialized nsize  in
    let ndata_tail = A.makeUninitialized nsize  in (* keep track of tail *)
    C.bucketsSet h  h_buckets;          (* so that indexfun sees the new bucket count *)
    for i = 0 to osize - 1 do
      insert_bucket ~hash ~h_buckets ~ndata_tail h (A.getUnsafe odata i)
    done;
    for i = 0 to nsize - 1 do
      match C.toOpt (A.getUnsafe ndata_tail i) with
      | None -> ()
      | Some tail -> N.nextSet tail C.emptyOpt
    done
  end


let add0 ~hash h key value =
  let h_buckets = C.buckets h in  
  let h_buckets_lenth = Array.length h_buckets in 
  let i = (Bs_Hash.getHashInternal hash) key [@bs] land (h_buckets_lenth - 1) in 
  let bucket = 
    N.bucket ~key ~value ~next:(A.getUnsafe h_buckets i) in  
  A.setUnsafe h_buckets i  (C.return bucket);
  let h_new_size = C.size h + 1 in 
  C.sizeSet h  h_new_size;
  if h_new_size > h_buckets_lenth lsl 1 then resize ~hash  h


let rec remove_bucket ~eq h h_buckets  i key prec buckets =
  match C.toOpt buckets with
  | None -> ()
  | Some cell ->
    let cell_next = N.next cell in 
    if (Bs_Hash.getEqInternal eq) (N.key cell) key [@bs]
    then 
      begin
        (match C.toOpt prec with
         | None -> A.setUnsafe h_buckets i  cell_next
         | Some c -> N.nextSet c cell_next);
        C.sizeSet h (C.size h - 1);        
      end
    else remove_bucket ~eq h h_buckets i key buckets cell_next

let remove0 ~hash ~eq h key =  
  let h_buckets = C.buckets h in 
  let i = (Bs_Hash.getHashInternal hash) key [@bs] land (Array.length h_buckets - 1) in  
  remove_bucket ~eq h h_buckets i key C.emptyOpt (A.getUnsafe h_buckets i)

let rec removeAllBuckets ~eq h h_buckets  i key prec buckets =
  match C.toOpt buckets with
  | None -> ()
  | Some cell ->
    let cell_next = N.next cell in 
    if (Bs_Hash.getEqInternal eq) (N.key cell) key [@bs]
    then 
      begin
        (match C.toOpt prec with
         | None -> A.setUnsafe h_buckets i  cell_next
         | Some c -> N.nextSet c cell_next);
        C.sizeSet h  (C.size h - 1);        
      end;
    removeAllBuckets ~eq h h_buckets i key buckets cell_next

let removeAll0 ~hash ~eq h key =
  let h_buckets = C.buckets h in 
  let i = (Bs_Hash.getHashInternal hash) key [@bs] land (Array.length h_buckets - 1) in  
  removeAllBuckets ~eq h h_buckets i key C.emptyOpt (A.getUnsafe h_buckets i)


let rec find_rec ~eq key buckets = 
  match C.toOpt buckets with 
  | None ->
    None
  | Some cell ->
    if (Bs_Hash.getEqInternal eq) key (N.key cell) [@bs] then Some (N.value cell)
    else find_rec ~eq key  (N.next cell)

let get0 ~hash ~eq h key =
  let h_buckets = C.buckets h in 
  let nid = (Bs_Hash.getHashInternal hash) key [@bs] land (Array.length h_buckets - 1) in 
  match C.toOpt @@ A.getUnsafe h_buckets nid with
  | None -> None
  | Some cell1  ->
    if (Bs_Hash.getEqInternal eq) key (N.key cell1) [@bs] then 
      Some  (N.value cell1)
    else
      match C.toOpt (N.next  cell1) with
      | None -> None
      | Some cell2 ->
        if (Bs_Hash.getEqInternal eq) key 
            (N.key cell2) [@bs] then 
          Some (N.value cell2) else
          match C.toOpt (N.next cell2) with
          | None -> None
          | Some cell3 ->
            if (Bs_Hash.getEqInternal eq) key 
                (N.key cell3) [@bs] then 
              Some (N.value cell3)
            else 
              find_rec ~eq key (N.next cell3)


let getAll0 ~hash ~eq h key =
  let rec find_in_bucket buckets = 
    match C.toOpt buckets with 
    | None ->
      []
    | Some cell ->
      if (Bs_Hash.getEqInternal eq) 
          (N.key cell) key [@bs]
      then (N.value cell) :: find_in_bucket (N.next cell)
      else find_in_bucket (N.next cell)  in
  let h_buckets = C.buckets h in     
  let nid = (Bs_Hash.getHashInternal hash) key [@bs] land (Array.length h_buckets - 1) in 
  find_in_bucket (A.getUnsafe h_buckets nid)

let rec replace_bucket ~eq  key info buckets = 
  match C.toOpt buckets with 
  | None ->
    true
  | Some cell ->
    if (Bs_Hash.getEqInternal eq) (N.key cell) key [@bs]
    then
      begin
        N.keySet cell key;
        N.valueSet cell info;
        false
      end
    else
      replace_bucket ~eq key info (N.next cell)

let replace0 ~hash ~eq  h key info =
  let h_buckets = C.buckets h in 
  let i = (Bs_Hash.getHashInternal hash) key [@bs] land (Array.length h_buckets - 1) in 
  let l = Array.unsafe_get h_buckets i in  
  if replace_bucket ~eq key info l then begin
    A.setUnsafe h_buckets i (C.return 
                                       (N.bucket ~key ~value:info ~next:l));
    C.sizeSet h (C.size  h + 1);
    if C.size h > Array.length (C.buckets h) lsl 1 then resize ~hash h
    (* TODO: duplicate bucklets ? *)
  end 

let rec mem_in_bucket ~eq key cell = 
    (Bs_Hash.getEqInternal eq) 
      (N.key cell) key [@bs] || 
      (match C.toOpt (N.next cell) with 
      | None -> false 
      | Some nextCell -> 
      mem_in_bucket ~eq key nextCell)

let has0 ~hash ~eq h key =
  let h_buckets = C.buckets h in 
  let nid = (Bs_Hash.getHashInternal hash) key [@bs] land (Array.length h_buckets - 1) in 
  let bucket = (A.getUnsafe h_buckets nid) in 
  match C.toOpt bucket with 
  | None -> false 
  | Some bucket -> 
    mem_in_bucket ~eq key bucket


let make = C.make

(*  Wrapper  *)
let make (type elt) (type id) initialize_size ~(dict : (elt, id) dict)  =
  let module M = (val dict) in 
  C.make initialize_size ~hash:M.hash ~eq:M.eq
    
let clear  = C.clear 

let size  = C.size 
let forEach  = N.forEach 
let reduce = N.reduce
let logStats  = N.logStats

let add  h key info  = 
  add0 ~hash:(C.hash h) h key info 

let remove h key = 
  remove0 ~hash:(C.hash h) ~eq:(C.eq h) h key 

let removeAll h key = 
  removeAll0 ~hash:(C.hash h) ~eq:(C.eq h) h key 

let get h key =           
  get0 ~hash:(C.hash h) ~eq:(C.eq h) h key 

let getAll h key =           
  getAll0 ~hash:(C.hash h) ~eq:(C.eq h) h key   

let replace h key info =
  replace0 ~hash:(C.hash h) ~eq:(C.eq h) h key info

let has h key =           
  has0 ~hash:(C.hash h) ~eq:(C.eq h) h key   

let keepMapInPlace  =  N.keepMapInPlace

File: lib/js/bs_HashMultiMap.js
'use strict';

var Bs_internalBuckets = require("./bs_internalBuckets.js");
var Bs_internalBucketsType = require("./bs_internalBucketsType.js");

function insert_bucket(hash, h_buckets, ndata_tail, _, _old_bucket) {
  while(true) {
    var old_bucket = _old_bucket;
    if (old_bucket !== undefined) {
      var nidx = hash(old_bucket.key) & (h_buckets.length - 1 | 0);
      var match = ndata_tail[nidx];
      if (match !== undefined) {
        match.next = old_bucket;
      } else {
        h_buckets[nidx] = old_bucket;
      }
      ndata_tail[nidx] = old_bucket;
      _old_bucket = old_bucket.next;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function resize(hash, h) {
  var odata = h.buckets;
  var osize = odata.length;
  var nsize = (osize << 1);
  if (nsize >= osize) {
    var h_buckets = new Array(nsize);
    var ndata_tail = new Array(nsize);
    h.buckets = h_buckets;
    for(var i = 0 ,i_finish = osize - 1 | 0; i <= i_finish; ++i){
      insert_bucket(hash, h_buckets, ndata_tail, h, odata[i]);
    }
    for(var i$1 = 0 ,i_finish$1 = nsize - 1 | 0; i$1 <= i_finish$1; ++i$1){
      var match = ndata_tail[i$1];
      if (match !== undefined) {
        match.next = Bs_internalBucketsType.emptyOpt;
      }
      
    }
    return /* () */0;
  } else {
    return 0;
  }
}

function replace_bucket(eq, key, info, _buckets) {
  while(true) {
    var buckets = _buckets;
    if (buckets !== undefined) {
      if (eq(buckets.key, key)) {
        buckets.key = key;
        buckets.value = info;
        return /* false */0;
      } else {
        _buckets = buckets.next;
        continue ;
        
      }
    } else {
      return /* true */1;
    }
  };
}

function make(initialize_size, dict) {
  return Bs_internalBucketsType.make(dict[/* hash */0], dict[/* eq */1], initialize_size);
}

function size(prim) {
  return prim.size;
}

function add(h, key, info) {
  var hash = h.hash;
  var h$1 = h;
  var key$1 = key;
  var value = info;
  var h_buckets = h$1.buckets;
  var h_buckets_lenth = h_buckets.length;
  var i = hash(key$1) & (h_buckets_lenth - 1 | 0);
  var bucket = {
    key: key$1,
    value: value,
    next: h_buckets[i]
  };
  h_buckets[i] = bucket;
  var h_new_size = h$1.size + 1 | 0;
  h$1.size = h_new_size;
  if (h_new_size > (h_buckets_lenth << 1)) {
    return resize(hash, h$1);
  } else {
    return 0;
  }
}

function remove(h, key) {
  var hash = h.hash;
  var eq = h.eq;
  var h$1 = h;
  var key$1 = key;
  var h_buckets = h$1.buckets;
  var i = hash(key$1) & (h_buckets.length - 1 | 0);
  var eq$1 = eq;
  var h$2 = h$1;
  var h_buckets$1 = h_buckets;
  var i$1 = i;
  var key$2 = key$1;
  var _prec = Bs_internalBucketsType.emptyOpt;
  var _buckets = h_buckets[i];
  while(true) {
    var buckets = _buckets;
    var prec = _prec;
    if (buckets !== undefined) {
      var cell_next = buckets.next;
      if (eq$1(buckets.key, key$2)) {
        if (prec !== undefined) {
          prec.next = cell_next;
        } else {
          h_buckets$1[i$1] = cell_next;
        }
        h$2.size = h$2.size - 1 | 0;
        return /* () */0;
      } else {
        _buckets = cell_next;
        _prec = buckets;
        continue ;
        
      }
    } else {
      return /* () */0;
    }
  };
}

function removeAll(h, key) {
  var hash = h.hash;
  var eq = h.eq;
  var h$1 = h;
  var key$1 = key;
  var h_buckets = h$1.buckets;
  var i = hash(key$1) & (h_buckets.length - 1 | 0);
  var eq$1 = eq;
  var h$2 = h$1;
  var h_buckets$1 = h_buckets;
  var i$1 = i;
  var key$2 = key$1;
  var _prec = Bs_internalBucketsType.emptyOpt;
  var _buckets = h_buckets[i];
  while(true) {
    var buckets = _buckets;
    var prec = _prec;
    if (buckets !== undefined) {
      var cell_next = buckets.next;
      if (eq$1(buckets.key, key$2)) {
        if (prec !== undefined) {
          prec.next = cell_next;
        } else {
          h_buckets$1[i$1] = cell_next;
        }
        h$2.size = h$2.size - 1 | 0;
      }
      _buckets = cell_next;
      _prec = buckets;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function get(h, key) {
  var hash = h.hash;
  var eq = h.eq;
  var h$1 = h;
  var key$1 = key;
  var h_buckets = h$1.buckets;
  var nid = hash(key$1) & (h_buckets.length - 1 | 0);
  var match = h_buckets[nid];
  if (match !== undefined) {
    if (eq(key$1, match.key)) {
      return /* Some */[match.value];
    } else {
      var match$1 = match.next;
      if (match$1 !== undefined) {
        if (eq(key$1, match$1.key)) {
          return /* Some */[match$1.value];
        } else {
          var match$2 = match$1.next;
          if (match$2 !== undefined) {
            if (eq(key$1, match$2.key)) {
              return /* Some */[match$2.value];
            } else {
              var eq$1 = eq;
              var key$2 = key$1;
              var _buckets = match$2.next;
              while(true) {
                var buckets = _buckets;
                if (buckets !== undefined) {
                  if (eq$1(key$2, buckets.key)) {
                    return /* Some */[buckets.value];
                  } else {
                    _buckets = buckets.next;
                    continue ;
                    
                  }
                } else {
                  return /* None */0;
                }
              };
            }
          } else {
            return /* None */0;
          }
        }
      } else {
        return /* None */0;
      }
    }
  } else {
    return /* None */0;
  }
}

function getAll(h, key) {
  var hash = h.hash;
  var eq = h.eq;
  var h$1 = h;
  var key$1 = key;
  var find_in_bucket = function (_buckets) {
    while(true) {
      var buckets = _buckets;
      if (buckets !== undefined) {
        if (eq(buckets.key, key$1)) {
          return /* :: */[
                  buckets.value,
                  find_in_bucket(buckets.next)
                ];
        } else {
          _buckets = buckets.next;
          continue ;
          
        }
      } else {
        return /* [] */0;
      }
    };
  };
  var h_buckets = h$1.buckets;
  var nid = hash(key$1) & (h_buckets.length - 1 | 0);
  return find_in_bucket(h_buckets[nid]);
}

function replace(h, key, info) {
  var hash = h.hash;
  var eq = h.eq;
  var h$1 = h;
  var key$1 = key;
  var info$1 = info;
  var h_buckets = h$1.buckets;
  var i = hash(key$1) & (h_buckets.length - 1 | 0);
  var l = h_buckets[i];
  if (replace_bucket(eq, key$1, info$1, l)) {
    h_buckets[i] = {
      key: key$1,
      value: info$1,
      next: l
    };
    h$1.size = h$1.size + 1 | 0;
    if (h$1.size > (h$1.buckets.length << 1)) {
      return resize(hash, h$1);
    } else {
      return 0;
    }
  } else {
    return 0;
  }
}

function has(h, key) {
  var hash = h.hash;
  var eq = h.eq;
  var h$1 = h;
  var key$1 = key;
  var h_buckets = h$1.buckets;
  var nid = hash(key$1) & (h_buckets.length - 1 | 0);
  var bucket = h_buckets[nid];
  if (bucket !== undefined) {
    var eq$1 = eq;
    var key$2 = key$1;
    var _cell = bucket;
    while(true) {
      var cell = _cell;
      if (eq$1(cell.key, key$2)) {
        return /* true */1;
      } else {
        var match = cell.next;
        if (match !== undefined) {
          _cell = match;
          continue ;
          
        } else {
          return /* false */0;
        }
      }
    };
  } else {
    return /* false */0;
  }
}

var clear = Bs_internalBucketsType.clear;

var forEach = Bs_internalBuckets.forEach;

var reduce = Bs_internalBuckets.reduce;

var keepMapInPlace = Bs_internalBuckets.keepMapInPlace;

var logStats = Bs_internalBuckets.logStats;

exports.make = make;
exports.clear = clear;
exports.add = add;
exports.get = get;
exports.getAll = getAll;
exports.has = has;
exports.remove = remove;
exports.removeAll = removeAll;
exports.replace = replace;
exports.forEach = forEach;
exports.reduce = reduce;
exports.keepMapInPlace = keepMapInPlace;
exports.size = size;
exports.logStats = logStats;
/* No side effect */
